New popup for multi selecting in the messages screen

matrix-amiel opened the pull request:
Hi.
Here new feature for messages screen, enables to create list of patients and send message to the list.
![image](https://user-images.githubusercontent.com/17809866/39469408-b4c39e94-4d40-11e8-9ca6-f381cd120e46.png)

Thanks
Amiel   

below added by bradymiller:
Up For Grabs demo for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Lambda_-_Up_For_Grabs_Demo

sjpadgett commented:
Just a passing comment and I haven't looked at code closely but for the Patient ID in dialog are you using pid or pubpid? Most user's only know the public patient id and may be confused by parameter using private pid.

matrix-amiel commented:
ID is pubpid, I have tried to develop this screen similar to the popup of regular add patient, in the add patient popup ID means pubpid and also in the new popup.

matrix-amiel commented:
Hi @bradymiller 
Now after the all Travis errors were fixed.. we will be very happy if you could review this as soon as possible..

Thanks.
Amiel

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
```
Needs the standard license/author/copyright header

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+// of the License, or (at your option) any later version.
+
+include_once('../../globals.php');
+include_once("$srcdir/patient.inc");
```
both lines above should be using require_once

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <?php Header::setupHeader(['jquery', 'bootstrap', 'select2', 'opener']); ?>
```
remove jquery, bootstrap since these are already included by default

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+        <?php
+        if (isset($_GET['patients'])) {
+            foreach ($results as $index => $result) {
+                echo '<tr id="row'.$index.'">' .
```
attr()

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+                        '<td>' . $result['phone_home'] . '</td>' .
+                        '<td>' . $result['ss'] . '</td>' .
+                        '<td>' . $result['DOB'] . '</td>' .
+                        '<td>' . $result['pubpid'] . '</td>' .
```
text() in 5 lines above

matrix-amiel replied:
I clean all the variables in the line 36. 
`$result=array_map('text', $result);`
This array is printed in the JS also, therefore I clean all the data in the php section.

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+                        '<td>' . $result['ss'] . '</td>' .
+                        '<td>' . $result['DOB'] . '</td>' .
+                        '<td>' . $result['pubpid'] . '</td>' .
+                        '<td><i class="fa fa-remove remove-patient" onclick="removePatient('.$index.')"></i></td>' .
```
attr()

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+<?php } ?>
+
+//Initial select2 library for auto completing using ajax
+$('#by-id, #by-name').select2({
```
top.restoreSession()

matrix-amiel replied:
Why this function is needed? haven't saw similar code in the another projects..
Is it for frames layout?

Where I need to put it? in the callback of the response of the ajax? here I haven't callback..

bradymiller replied:
Place it before the ajax call. This prevents session leaking (when have multiple tabs of openemr in the same browser) and is discussed here: https://www.open-emr.org/wiki/index.php/OpenEMR_System_Architecture#PHP_Sessions_and_Browser_Windows

matrix-amiel replied:
Ok. thank you.

bradymiller replied:
missing the top.restoreSession() here

matrix-amiel replied:
I putted it into the ajax json beforeSend (line 170) (I can't change select2 library)

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+});
+
+//get all the data of selected patient
+$('#by-id').on('change', function () {
```
top.restoreSession()'

matrix-amiel replied:
Done

bradymiller replied:
move it to before the ajax call

matrix-amiel replied:
done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+});
+
+//get all the data of selected patient
+$('#by-name').on('change', function () {
```
top.restoreSession()'

matrix-amiel replied:
Done

bradymiller replied:
move it to before the ajax call

matrix-amiel replied:
done

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+});
+
+//add new patient to list
+$('#add-to-list').on('click', function (e) {
```
top.restoreSession()

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,258 @@
+//send array of patients to function 'setMultiPatients' of the opener
+function selPatients() {
+    if (opener.closed || ! opener.setMultiPatients)
+        alert("<?php echo htmlspecialchars(xl('The destination form was closed; I cannot act on your selection.'), ENT_QUOTES); ?>");
```
remove htmlspecialchar and xl here
and use xls() instead

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder_ajax.php:
```diff
@@ -0,0 +1,42 @@
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
```
Needs the standard license/author/copyright header

matrix-amiel replied:
Done

bradymiller commented on interface/main/finder/multi_patients_finder_ajax.php:
```diff
@@ -0,0 +1,42 @@
+// of the License, or (at your option) any later version.
+
+include_once('../../globals.php');
+include_once("$srcdir/patient.inc");
```
both lines above should be using require_once

matrix-amiel replied:
Done

bradymiller commented:
Placed a code review. Will test it and set up a up for grabs demo on next version.

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,275 @@
+    foreach ($patients as $patient) {
+        $result=getPatientData($patient, 'id, pid, lname, fname, mname, pubpid, ss, DOB, phone_home');
+        //clean data using 'text' function
+        $result=array_map('text', $result);
```
don't do this since it's messy and you never know where the data is going to be used in the future:
1. note text() is for html escaping data
2. if the data is used in algorithms it can get messed up
3. note it needs to be done after the oeFormatShortDate call

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,275 @@
+        $result=getPatientData($patient, 'id, pid, lname, fname, mname, pubpid, ss, DOB, phone_home');
+        //clean data using 'text' function
+        $result=array_map('text', $result);
+        $result['DOB'] = oeFormatShortDate($result['DOB']);
```
Also best to only call oeFormatShortDate when it is outputted since any algorithms within script should use he YYYY-MM--DD format.

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,275 @@
+                        '<td>' . $result['phone_home'] . '</td>' .
+                        '<td>' . $result['ss'] . '</td>' .
+                        '<td>' . $result['DOB'] . '</td>' .
+                        '<td>' . $result['pubpid'] . '</td>' .
```
see comments above

bradymiller commented:
Up For Grabs demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Lambda_-_Up_For_Grabs_Demo

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,273 @@
+ * @package OpenEMR
+ * @author  Amiel Elboim <amielel@matrix.co.il>
+ * @link    http://www.open-emr.org
+ */
```
Instead, use the new briefer format for the above header. Here is an example:
https://www.open-emr.org/wiki/index.php/How_to_Document_Your_Code_Properly#File-Level_DocBlock

bradymiller commented on interface/main/finder/multi_patients_finder_ajax.php:
```diff
@@ -0,0 +1,57 @@
+ * @package OpenEMR
+ * @author  Amiel Elboim <amielel@matrix.co.il>
+ * @link    http://www.open-emr.org
+ */
```
Instead, use the new briefer format for the above header. Here is an example:
https://www.open-emr.org/wiki/index.php/How_to_Document_Your_Code_Properly#File-Level_DocBlock

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,273 @@
+            <div class="select-box">
+                <label><?php echo xlt('Patient name') .':'; ?></label>
+                <select id="by-name" class="input-sm">
+                    <option value=""><?php echo xlt('Type name'); ?></option>
```
Rather then Type, use Enter

bradymiller commented on interface/main/finder/multi_patients_finder.php:
```diff
@@ -0,0 +1,273 @@
+                </select>
+                <label><?php echo xlt('Patient ID'); ?></label>
+                <select id="by-id" class="input-sm">
+                    <option value=""><?php echo xlt('Type ID'); ?></option>
```
Rather then Type, use Enter

bradymiller commented:
Review complete. Looking forward to the next revision, which I am guessing will be ready for the codebase :)

matrix-amiel commented:
Hi.
I fixed according your review.
Regarding the bugs I fixed bug 2, bug 1 I didn't succeed to reproduce but I think it happen because the first patient not removed (bug 2). but please check it again. 

Thank you for a quick reviewing!

bradymiller commented:
Testing very well. Will bring this into codebase. Thanks for the improvement! -brady
